### ISSUE TITLE
Windows: Can bind socket on IP not used by machine. 

### ISSUE THREAD

**Issue by mascojames** [nanomsg/nanomsg#796]
Nanomsg on Windows allows for binding on IP address not belonging to the current machine. A failure does not occur til further in the process. This does not occur on Linux, Linux fails at binding the socket, Windows should behave similarly.

Here's a small easily repeatable test case (In C++11):
```
   int outgoing_socket = nn_socket (AF_SP, NN_PUB);

   if (outgoing_socket < 0)
   {
      std::cout << "Failed to create socket: " << nn_strerror(nn_errno()) << std::endl;
   }
   else
   {
      std::string topic = "tcp://123.45.67.89:12345";
      std::this_thread::sleep_for(std::chrono::milliseconds(100));
      int endpoint_id = nn_bind (outgoing_socket, topic.c_str());
      if(endpoint_id < 0)
      {
         std::cout << "Failed to bind socket: " << nn_strerror(nn_errno()) << std::endl ;
      }
      else
      {
         std::cout << "Connected to: " << topic << std::endl;
         char buff[10] = {0};
         std::this_thread::sleep_for(std::chrono::milliseconds(100));
         int bytes_sent = nn_send(outgoing_socket, buff, 10, 0);
         if (bytes_sent == 10)
         {
            std::cout << "Successfully sent 10 bytes" << std::endl;
         }
         else
         {
            std::cout << "Failed to send bytes" << std::endl;
         }
         std::this_thread::sleep_for(std::chrono::milliseconds(100));
         nn_shutdown(outgoing_socket, endpoint_id);

         std::cout << "Shutdown Complete.";
      }
      std::this_thread::sleep_for(std::chrono::milliseconds(100));
      nn_close(outgoing_socket);
      std::cout << "Close Complete.";
   }
```

Output is:
````
1>  Connected to: tcp://123.45.67.89:12345
1>  Successfully sent 10 bytes
1>  Assertion failed: 3 == NN_TIMER_STATE_IDLE (..\..\src\aio\timer.c:62)
````

**Comment by gdamore** [comment]
Bizarre!  Ok, I'll take a look at it.

**Comment by gdamore** [comment]
This is quite strange, in that it shouldn't matter if the socket is actually bound or not -- I'd expect the underlying send of data to stall or fail, just as if a remote peer was stuck.  I need to fire up windows with a debugger because the problem isn't obvious -- the Windows behavior, while unfortunate, should actually work (except that no message would be delivered of course.)

**Comment by gdamore** [comment]
I think I have it figured out.  We need to unconditionally stop the streamhdr timer.

**Comment by gdamore** [comment]
Actually this is even more stupid.
We have an extra state in the timer logic, and yes, the stream hdr timer is indeed still running.  The entire timer framework is buggy here, since we can wind up racing against the worker_execute() of the magic stop task.

The thing is, we don't need that extra state or the other level of asynchronous behavior.  Nothing uses it.  It only serves to obfuscate and create an opportunity for a race condition.  Better to just cancel the timeout synchronously!

**Comment by gdamore** [comment]
Actually I was sadly wrong.  FSM for the loss again.  Stay tuned.

**Comment by gdamore** [comment]
Can you confirm that you are using the current nanomsg version (1.0.0?)

**Comment by gdamore** [comment]
So I made a test case to try to reproduce this on Windows (the assertion failure), but I cannot:

https://github.com/nanomsg/nanomsg/pull/801

The windows runs on AppVeyor are all clean:

https://ci.appveyor.com/project/nanomsg/nanomsg/build/0.8.249

So I'm not sure what the problem is?  The fact that nn_bind() didn't fail?  That's fine -- and there are sound reasons why Windows may have chosen this behavior, and shouldn't affect us.   (There is a moral of the story here, which is that you ought not assume nanomsg is checking IP addresses...)

**Comment by mascojames** [comment]
Yes I am using the released 1.0.0 version of nanomsg. I guess that is acceptable that Windows and Linux behave differently in that aspect (nn_bind).
